File: src/main/java/com/nerdearla/workshop/user/buyer/client/BuyerClient.java
package com.nerdearla.workshop.user.buyer.client;

import com.nerdearla.workshop.user.buyer.Buyer;
import com.nerdearla.workshop.user.buyer.error.BuyerRetrievingError;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Component;
import org.springframework.web.reactive.function.client.ClientResponse;
import org.springframework.web.reactive.function.client.WebClient;
import reactor.core.publisher.Mono;

@Component
public class BuyerClient {

    private static final Logger LOGGER = LoggerFactory.getLogger(BuyerClient.class);
    private final WebClient webClient;

    public BuyerClient(WebClient webClient) {
        this.webClient = webClient;
    }

    public Buyer getById(String id) {
        Buyer buyer = webClient
                .get()
                .uri("/buyers/{id}", id)
                .retrieve()
                .onStatus(HttpStatus::isError, this::handleError)
                .bodyToMono(Buyer.class)
                .block();
        LOGGER.info("buyer found: {}", buyer);
        return buyer;
    }

    private Mono<Throwable> handleError(ClientResponse response) {
        LOGGER.error("Error while communicating with buyer service, {}", response);
        return Mono.error(new BuyerRetrievingError());
    }
}

File: src/main/java/com/nerdearla/workshop/user/buyer/client/BuyerClient.kt
package com.nerdearla.workshop.user.buyer.client

import com.nerdearla.workshop.shared.client.get
import com.nerdearla.workshop.shared.utils.CompanionLogger
import com.nerdearla.workshop.user.buyer.Buyer
import com.nerdearla.workshop.user.buyer.error.BuyerRetrievingError
import org.springframework.stereotype.Component
import org.springframework.web.reactive.function.client.ClientResponse
import org.springframework.web.reactive.function.client.WebClient
import reactor.core.publisher.Mono

@Component
class BuyerClient(
    private val client: WebClient
) {

    fun getBy(id: String): Buyer =
        client.get<Buyer>("/buyers/{buyerId}", id, handler = ::mono)
            .log { info("buyer found: {}", it) }

    private fun mono(response: ClientResponse) = Mono.error<Throwable>(BuyerRetrievingError())
        .log { error("Error while communicating with buyer service, {}", response) }

    companion object : CompanionLogger()
}
